openstack/identity/v3/credentials/requests.go

```diff
@@ -0,0 +1,123 @@
+// the List request
+type ListOptsBuilder interface {
+	ToCredentialListQuery() (string, error)
+}
+
+// ListOpts provides options to filter the List results.
+type ListOpts struct {
+	// UserID filters the response by a credential user_id
+	UserID string `json:"user_id"`
+}
+
+// ToCredentialListQuery formats a ListOpts into a query string.
+func (opts ListOpts) ToCredentialListQuery() (string, error) {
+	q, err := gophercloud.BuildQueryString(opts)
+	return q.String(), err
+}
+
+// List enumerates the Credentials to which the current token has access.
+func List(client *gophercloud.ServiceClient, opts ListOptsBuilder) pagination.Pager {
+	url := listURL(client)
+	if opts != nil {
+		query, err := opts.ToCredentialListQuery()
+		if err != nil {
+			return pagination.Pager{Err: err}
+		}
+		url += query
+	}
+	return pagination.NewPager(client, url, func(r pagination.PageResult) pagination.Page {
+		return CredentialPage{pagination.LinkedPageBase{PageResult: r}}
+	})
+}
+
+// Get retrieves details on a single user, by ID.
+func Get(client *gophercloud.ServiceClient, id string) (r GetResult) {
+	_, r.Err = client.Get(getURL(client, id), &r.Body, nil)
+	return
+}
+
+// CreateOptsBuilder allows extensions to add additional parameters to
+// the Create request.
+type CreateOptsBuilder interface {
+	ToCredentialCreateMap() (map[string]interface{}, error)
+}
+
+// CreateOpts provides options used to create a credential.
+type CreateOpts struct {
+	// Serialized blob containing the credentials
+	Blob string `json:"blob" required:"true"`
+	// ID of the project.
+	ProjectID string `json:"project_id" required:"true"`
+	// The type of the credential.
+	Type string `json:"type" required:"true"`
+	// ID of the user who owns the credential.
+	UserID string `json:"user_id" required:"true"`
+}
+
+// ToCredentialCreateMap formats a CreateOpts into a create request.
+func (opts CreateOpts) ToCredentialCreateMap() (map[string]interface{}, error) {
+	return gophercloud.BuildRequestBody(opts, "credential")
+}
+
+// Create creates a new Credential.
+func Create(client *gophercloud.ServiceClient, opts CreateOptsBuilder) (r CreateResult) {
+	b, err := opts.ToCredentialCreateMap()
+	if err != nil {
+		r.Err = err
+		return
+	}
+	_, r.Err = client.Post(createURL(client), &b, &r.Body, &gophercloud.RequestOpts{
+		OkCodes: []int{201},
+	})
+	return
+}
+
+// Delete deletes a credential.
+func Delete(client *gophercloud.ServiceClient, id string) (r DeleteResult) {
+	_, r.Err = client.Delete(deleteURL(client, id), nil)
+	return
+}
+
+// UpdateOptsBuilder allows extensions to add additional parameters to
+// the Update request.
+type UpdateOptsBuilder interface {
+	ToCredentialsUpdateMap() (map[string]interface{}, error)
+}
+
+// UpdateOpts represents parameters to update a credential.
+type UpdateOpts struct {
+	// Serialized blob containing the credentials.
+	Blob string `json:"blob,omitempty"`
+	// ID of the project.
+	ProjectID string `json:"project_id" required:"true"`
```

Comment:
`project_id` is not required, when the type is not `ec2`. https://github.com/openstack/keystone/blob/00663f93b7e9513602d73dcd0aba7ad1cc1feefc/keystone/credential/schema.py#L44